### PR TITLE
Move sha1 digest calculation to native function

### PR DESCRIPTION
**Pull request by FooIbar**
`kotlinx-io` has removed hash functions API.

### CODE REVIEW

File: app/src/main/cpp/hash.c [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        hex[2 * i + 1] = hex_digits[byte & 0xF];
+    }
+    hex[2 * length] = '\0';
+    return hex;
+}
+
+JNIEXPORT jstring JNICALL
+Java_com_hippo_ehviewer_jni_HashKt_sha1(JNIEnv *env, jclass clazz, jint fd) {
+    EH_UNUSED(clazz);
+    byte *buffer = malloc(BUFFER_SIZE);
```
Comment (VDostoyevskiy) [inline app/src/main/cpp/hash.c:35]: Allocate buffer on stack

File: app/src/main/cpp/hash.c [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    size_t bytes_read;
+    while ((bytes_read = read(fd, buffer, BUFFER_SIZE)) > 0) {
+        sha1_update(&ctx, bytes_read, buffer);
+    }
+
+    byte digest[SHA1_DIGEST_SIZE];
+    sha1_digest(&ctx, SHA1_DIGEST_SIZE, digest);
+
+    char *hex_digest = bytes_to_hex(digest, SHA1_DIGEST_SIZE);
```
Comment (VDostoyevskiy) [inline app/src/main/cpp/hash.c:52]: Allocate hex on stack and pass it

File: app/src/main/cpp/hash.c (hunk 1 above)
Comment (FooIbar) [reply app/src/main/cpp/hash.c:35]: Should inline `bytes_to_hex` as well I guess, it most likely won't be used anywhere else.

**Comment by VDostoyevskiy** [comment]
Btw I suggest write it in rust as we haven't compile nettle with simd support

**Comment by VDostoyevskiy** [comment]
> Btw I suggest write it in rust as we haven't compile nettle with simd support

https://github.com/RustCrypto/hashes

**Review by VDostoyevskiy — APPROVED** [review]

**Comment by VDostoyevskiy** [comment]
> > we haven't compile nettle with simd support
> 
> Is anything preventing us from doing that?

IDK if llvm's integrated assembler can compile nettle's assembly, maybe or maybe not

**Pull request merged**